hs/app/reachpc/ReachPC/DirScan.hs

```diff
@@ -0,0 +1,108 @@
+ where
+  readHash (p, fh) = case fh of
+    FH_File _ -> sequence (p, FD_File <$> BS.readFile (root </> p))
+    FH_Directory -> return (p, FD_Directory)
+    FH_Deleted -> return (p, FD_Deleted)
+
+-- Produces a hash tree of file updates that need to be sent to the remote.
+-- Removes redundant hashes and records deleted files
+compareHashTrees :: FileHashTree -> FileHashTree -> FileHashTree
+compareHashTrees localHashes remoteHashes = M.union updatedFiles deletedFiles
+ where
+  deletedFiles = M.map (const FH_Deleted) $ M.difference localHashes remoteHashes
+  updatedFiles = M.differenceWith hasChanged remoteHashes localHashes
+  hasChanged l r = if l == r then Nothing else Just l
+
+hashFile :: FilePath -> IO Hash
+hashFile = fmap hash . BS.readFile
+
+-- Takes a directory and walks it recursively. Hashes each file encountered,
+-- records directories encountered, and ignores symlinks. Special files ???
+-- Reads ".reachignore" if it exists and ignores files that match patterns in it
+hashDirectory :: FilePath -> IO FileHashTree
+hashDirectory root = M.mapKeys dropRootPrefix <$> doScan [] [root] M.empty
+ where
+  dropRootPrefix = drop $ length $ addTrailingPathSeparator root
+  isIgnored ignores path = any (`G.match` path) ignores
+  readIgnores dir = do
+    let path = dir </> ".reachignore"
+    haveIgnores <- doesFileExist path
+    case haveIgnores of
+      False -> return []
+      True -> do
+        ignoresFile <- readFile path
+        let patternChar c = not $ isSpace c || c == '#'
+        let ignoresPatterns = filter (not . null) $ map (takeWhile patternChar) $ lines ignoresFile
+        let ignores = map (G.compile . (dir </>)) ignoresPatterns
+        return ignores
+  doScan _ [] hashTree = return hashTree
+  doScan ignores (dir:dirStack) hashTree = do
+    -- Read .reachignore in this directory
+    newIgnores <- readIgnores dir
+    let ignores' = newIgnores <> ignores
+    
+    -- Walk current dir
+    paths <- listDirectory dir
+    let paths' = map (dir </>) paths
+    pathsAreFile <- mapM doesFileExist paths'
+    pathsAreSymlink <- mapM pathIsSymbolicLink paths'
+    let pathTypes = zip3 paths' pathsAreFile pathsAreSymlink
+    let pathTypes' = [(p, isFile) | (p, isFile, isSymlink) <- pathTypes, 
+                                    not isSymlink, 
+                                    not $ isIgnored ignores' p]
+    let (files, dirs) = partition snd pathTypes'
+    let files' = map fst files
+    let dirs' = map fst dirs
+    
+    -- Hash encountered files
+    fileHashes <- mapM hashFile files'
+    let fileHashes' = zip files' (map FH_File fileHashes)
+    let newFilesMap = M.fromList fileHashes'
+
+    -- Add remaining dirs to stack and hashes map
+    let dirStack' = dirs' <> dirStack
+    let newDirsMap = M.fromList (map (, FH_Directory) dirs')
+    
+    let hashTree' = M.unions [hashTree, newFilesMap, newDirsMap]
+    doScan ignores' dirStack' hashTree'
```

Comment:
This makes it so the ignores in `/a/b/.reachignore` affect `/a/c/` if the stack is `/a/b : /a/c : []`

Reply:
this is not true. 
![image](https://user-images.githubusercontent.com/1377477/174152693-66900eff-6085-4bff-8e9b-f54a241727da.png)

Reply:
If `/a/b/.reachignore` contains `*.js` but `/a/c/index.js` is a file and you're going to visit `/a/c` after you visit `/a/b`, then `ignores'` is going to have `*.js`, right?

Reply:
If it doesn't, then it seems weird, because I WANT `/a/b/xyz/index.js` to be ignored... right? If it doesn't matter, then you're making `ignores'` get bigger and bigger and bigger rather than scoping it. I basically don't get why this is a queue rather than being a tree walk

Reply:
The filesystem is already inherently scoped. It's less efficient in that it in a big project it will check a lot of impossible patterns, but I think it's safe to assume it will be super rare for that to be noticable.